lib/mongoid/criteria/queryable/selector.rb

```diff
@@ -184,16 +200,71 @@ def evolve_hash(serializer, value)
           end
         end
 
-        # Evolve a single key selection with range values.
+        # Evolve a single key selection with range values. This method traverses
+        # the association tree to build a query for the given value and
+        # serializer. There are three parts to the query here:
         #
-        # @param [ Object ] serializer The optional serializer for the field.
-        # @param [ Range ] value The Range to serialize.
+        # (1) "klass.child.gchild" => {
+        #       "$elemMatch" => {
+        #     (2) "ggchild.field" => (3) { "$gte" => 6, "$lte" => 10 }
+        #       }
+        #     }
+        # (1) The first n fields are dotted together until the last
+        #     embeds_many or field of type array. In the above case, gchild
+        #     would be an embeds_many or Array, and ggchild would be an
+        #     embeds_one or a hash.
+        # (2) The last fields are used inside the $elemMatch. This one is
+        #     actually optional, and will be ignored if the last field is an
+        #     array or embeds_many. If the last field is an array (1), (2) and
+        #     (3) will look like:
+        #
+        #       "klass.child.gchild.ggchild.field" => {
+        #         { "$elemMatch" => { "$gte" => 6, "$lte" => 10 } }
+        #       }
+        #
+        # (3) This is calculated by:
         #
-        # @return [ Range ] The serialized Range.
+        #       value.__evolve_range__(serializer: serializer).
         #
         # @api private
-        def evolve_range(serializer, value)
-          value.__evolve_range__(serializer: serializer)
+        #
+        # @param [ String ] key The to store the range for.
+        # @param [ Object ] serializer The optional serializer for the field.
+        # @param [ Range ] value The Range to serialize.
+        #
+        # @return [ Array<String, Hash> ] The store name and serialized Range.
+        def evolve_range(key, serializer, value)
+          v = value.__evolve_range__(serializer: serializer)
+          rels = []
```

Comment:
While there are many places in Mongoid that use the word "relation", the correct term is "association" and all new code should use the latter.

Reply:
done